Parquet-313: Implement 3 level list writing rule for Parquet-Thrift

SinghAsDev opened the pull request:
Parquet-thrift does not write three level lists, unlike other object models in the project. This JIRA should add 3 level list writing rule's implementation for parquet-thrift.

rdblue commented:
@SinghAsDev can you look into the test failures and get this branch passing?

SinghAsDev commented:
@rdblue this should be fixed now.

SinghAsDev commented:
The build with hadoop-2 profile failed. Not sure why. I tried locally and it passes.

rdblue commented:
@SinghAsDev, the error messages from the Travis logs are:

```
Tests in error: 
  testThriftOptionalFieldsWithReadProjectionUsingParquetSchema(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testReorderdOptionalFields(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testProjectOutOptionalFields(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testPullInRequiredMaps(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testDropMapValuePrimitive(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testDropMapValueStruct(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testDropMapValueNestedPrim(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testDropMapValueNestedStruct(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testPullInRequiredLists(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testPullInRequiredSets(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testPullInPrimitiveValues(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testPullingInRequiredStructWithFilter(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFile(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteStatistics(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFileListOfMap(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFileMapOfList(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFileMapOfLists(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFileWithThreeLevelsList(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFileListOfMapWithThreeLevelLists(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
```

`TaskAttemptContext` was changed from a class to an interface in Hadoop 2. That's a binary incompatible change, so we have to use shims, in `ContextUtil`, to dynamically resolve the right method. You're probably getting the configuration from the context somewhere and need to use the helper from `ContextUtil`.

SinghAsDev commented:
Hey @rdblue would it be possible for you to review this?

piyushnarang commented on parquet-scrooge/src/main/java/org/apache/parquet/scrooge/ScroogeReadSupport.java:
```diff
@@ -37,8 +38,14 @@ public ScroogeReadSupport() {
   protected MessageType getProjectedSchema(FieldProjectionFilter fieldProjectionFilter) {
     ThriftType.StructType thriftStruct = new ScroogeStructConverter().convert(thriftClass);
-    return new ThriftSchemaConverter(fieldProjectionFilter).convert(thriftStruct);
+    return getProjectedSchema(null, fieldProjectionFilter);
```
maybe pass an empty config rather than a null object?

piyushnarang commented on parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ParquetThriftBytesOutputFormat.java:
```diff
@@ -57,13 +58,33 @@ public ParquetThriftBytesOutputFormat() {
+                                        Class<? extends TBase<?, ?>> thriftClass,
+                                        boolean buffered,
+                                        FieldIgnoredHandler errorHandler) {
+    this(null, protocolFactory, thriftClass, buffered, errorHandler);
```
empty config instead of null?

piyushnarang commented on parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java:
```diff
@@ -79,14 +79,34 @@
+                                 Class<? extends TBase<?, ?>> thriftClass,
+                                 boolean buffered,
+                                 FieldIgnoredHandler errorHandler) {
+    this(null, protocolFactory, thriftClass, buffered, errorHandler);
```
here too avoid null

piyushnarang commented on parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java:
```diff
@@ -117,8 +138,9 @@ public WriteContext init(Configuration configuration) {
-    this.thriftStruct = ThriftSchemaConverter.toStructType(thriftClass);
-    this.schema = ThriftSchemaConverter.convertWithoutProjection(thriftStruct);
+    ThriftSchemaConverter thriftSchemaConverter = new ThriftSchemaConverter(this.configuration);
+    this.thriftStruct = thriftSchemaConverter.toStructType(thriftClass);
```
can drop the `this.` for this line and the next right?

piyushnarang commented on parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java:
```diff
@@ -158,7 +180,8 @@ Currently this fix only applies to TBinaryProtocol which has the setReadLength d
   public void prepareForWrite(RecordConsumer recordConsumer) {
     final MessageColumnIO columnIO = new ColumnIOFactory().getColumnIO(schema);
-    this.parquetWriteProtocol = new ParquetWriteProtocol(recordConsumer, columnIO, thriftStruct);
+    this.parquetWriteProtocol = new ParquetWriteProtocol(
```
here too can drop the `this` for parquetWriteProtoCol and configuration.

piyushnarang commented on parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftReadSupport.java:
```diff
@@ -189,8 +189,18 @@ public ThriftReadSupport(Class<T> thriftClass) {
+  @SuppressWarnings("unchecked")
+  protected MessageType getProjectedSchema(FieldProjectionFilter
+    fieldProjectionFilter) {
+    return new ThriftSchemaConverter(null, fieldProjectionFilter)
```
empty instead of null?

piyushnarang commented on parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java:
```diff
@@ -98,13 +104,46 @@ public void writeI32(int i32) throws TException {
+
+  }
+
+  class TwoLevelListWriteProtocol extends ListWriteProtocol{
```
nit: add a space after `Protocol` and before `{`

piyushnarang commented on parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java:
```diff
@@ -133,36 +173,67 @@ private void startListWrapper() {
-      size = list.size;
-      startListWrapper();
-    }
+  class ThreeLevelListWriteProtocol extends ListWriteProtocol{
```
nit: add a space after `Protocol` and before `{`

piyushnarang commented on parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java:
```diff
@@ -677,10 +758,22 @@ private TProtocol getProtocol(ThriftField field, ColumnIO columnIO, Events retur
       break;
     case SET:
-      p = new ListWriteProtocol((GroupColumnIO)columnIO, ((SetType)type).getValues(), returnClause);
+      if (writeThreeLevelList) {
```
we could refactor this out into a method `createListWriteProtocol(writeLevel3, columnIO, type, returnClause)` that way we don't have this largely identical repeated block

piyushnarang commented on parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java:
```diff
@@ -76,21 +78,36 @@
+
+  private ThriftSchemaConvertVisitor(FieldProjectionFilter fieldProjectionFilter, boolean doProjection,
+                                     boolean keepOneOfEachUnion) {
+    this(fieldProjectionFilter, doProjection, keepOneOfEachUnion, null);
```
avoid null?

SinghAsDev commented:
@piyushnarang your feedbacks should be addressed with latest change.

SinghAsDev commented:
@rdblue / @isnotinvain I would highly appreciate if you guys can review this and help in merging this. This will avoid us from creating an internal fork.

nandorKollar commented on parquet-scrooge/src/main/java/org/apache/parquet/scrooge/ScroogeReadSupport.java:
```diff
@@ -37,8 +38,14 @@ public ScroogeReadSupport() {
+
+  @Deprecated
   protected MessageType getProjectedSchema(FieldProjectionFilter fieldProjectionFilter) {
     ThriftType.StructType thriftStruct = new ScroogeStructConverter().convert(thriftClass);
```
thriftStruct is an unused variable, the result of this conversion is not used, could you please delete this?

SinghAsDev replied:
My bad, this is now removed.

julienledem commented:
@ttim and @tlazaro, do you mind taking a look and letting us know if this looks ok to you? We'd want to get this merged.

SinghAsDev commented:
> @ttim and @tlazaro, do you mind taking a look and letting us know if this looks ok to you? We'd want to get this merged.

Hey @ttim and @tlazaro, we have been using this at Pinterest for a few years just fine, and are wanting to have this merged in so that we can get rid of our internal fork. Also, this is config controlled, so default behavior remains unchanged. Would really appreciate if you can help reviewing this from twitter side, so that we can go ahead with merging this. I can rebase the patch right after.

gszadovszky commented:
@SinghAsDev, could you please rebase this PR to have the conflicts resolved?

SinghAsDev commented:
> @SinghAsDev, could you please rebase this PR to have the conflicts resolved?

Hey @gszadovszky this is now rebased, let me know if there is anything else I can do to have this merged, thanks!

tlazaro reviewed (approved):
Thanks for your work and for bringing our attention to it!

I don't have approving rights but I think this is fine. Timur will be able to check this sometime next week.

I have one question though, I would like to understand better. Couldn't we have done a version that works for any level of nesting of lists? I wouldn't want to see a 'need to support 4 level lists' in the future. Not grounds to block this PR, we can always revisit if needed and given there will be a config flag from now on, easy to enhance later.

SinghAsDev commented:
> Thanks for your work and for bringing our attention to it!
> 
> I don't have approving rights but I think this is fine. Timur will be able to check this sometime next week.
> 
> I have one question though, I would like to understand better. Couldn't we have done a version that works for any level of nesting of lists? I wouldn't want to see a 'need to support 4 level lists' in the future. Not grounds to block this PR, we can always revisit if needed and given there will be a config flag from now on, easy to enhance later.

Thanks for the review!

I actually don't see a reason for having more level of nesting. The three level nesting was needed to support nulls at all levels of list definition. Moreover, this is how other models have it implemented as well.

gszadovszky commented on parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ParquetThriftBytesOutputFormat.java:
```diff
@@ -57,13 +58,33 @@ public ParquetThriftBytesOutputFormat() {
+        configuration, protocolFactory, thriftClass, buffered, errorHandler));
+  }
+
+  @Deprecated
+  /**
+   * @deprecated Use @link{ParquetThriftBytesOutputFormat(
+   * Configuration configuration, TProtocolFactory protocolFactory,
+   * Class<? extends TBase<?, ?>> thriftClass, boolean buffered,
+   * FieldIgnoredHandler errorHandler)} instead
+   */
+  public ParquetThriftBytesOutputFormat(TProtocolFactory protocolFactory,
```
nit:
```suggestion
  /**
   * Configuration configuration, TProtocolFactory protocolFactory,
   * Class<? extends TBase<?, ?>> thriftClass, boolean buffered,
   * FieldIgnoredHandler errorHandler)} instead
   * @deprecated Use @link{ParquetThriftBytesOutputFormat(
   */
  @Deprecated
  public ParquetThriftBytesOutputFormat(TProtocolFactory protocolFactory,
```

gszadovszky commented on parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java:
```diff
@@ -79,14 +79,34 @@
   }
 
-  public ThriftBytesWriteSupport(TProtocolFactory protocolFactory, Class<? extends TBase<?, ?>> thriftClass, boolean buffered, FieldIgnoredHandler errorHandler) {
+  @Deprecated
+  /**
+   * @deprecated Use @link{ThriftBytesWriteSupport(Configuration configuration,
+   * TProtocolFactory protocolFactory, Class<? extends TBase<?, ?>> thriftClass,
+   * boolean buffered, FieldIgnoredHandler errorHandler)} instead
+   */
+  public ThriftBytesWriteSupport(TProtocolFactory protocolFactory,
```
nit:
```suggestion
  /**
   * TProtocolFactory protocolFactory, Class<? extends TBase<?, ?>> thriftClass,
   * boolean buffered, FieldIgnoredHandler errorHandler)} instead
   * @deprecated Use @link{ThriftBytesWriteSupport(Configuration configuration,
   */
  @Deprecated
  public ThriftBytesWriteSupport(TProtocolFactory protocolFactory,
```

gszadovszky commented on parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java:
```diff
@@ -51,6 +52,11 @@
 
 public class ParquetWriteProtocol extends ParquetProtocol {
 
+  public static final String WRITE_THREE_LEVEL_LISTS = "parquet.thrift.write-three-level-lists";
```
I know it is only about this change but we always struggles with our documentation. We have started to document the available configs in [parquet-hadoop](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/README.md) and also in [parquet-avro](https://github.com/apache/parquet-mr/blob/master/parquet-avro/README.md). Maybe, it worth a similar documentation in parquet-thrift as well.

SinghAsDev replied:
Good point, I added a Reademe with info on config this diff adds. I can follow up with a separate diff to add more info on the Readme.

gszadovszky commented on parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java:
```diff
@@ -76,22 +78,37 @@
   private final boolean keepOneOfEachUnion;
 
-  private ThriftSchemaConvertVisitor(FieldProjectionFilter fieldProjectionFilter, boolean doProjection, boolean keepOneOfEachUnion) {
+  private boolean writeThreeLevelList = ParquetWriteProtocol.WRITE_THREE_LEVEL_LISTS_DEFAULT;
```
It seems this variable gets a value in the constructor anyway. I would suggest having it final and assign value in the constructor only once.

SinghAsDev replied:
Good point, made the change.

gszadovszky reviewed (changes requested):
I've run through the changeset and found some minor issues.
I am not an expert in parquet-thrift so let's wait for @ttim's approval. I'll push this change if there are no objections.

SinghAsDev commented:
> I've run through the changeset and found some minor issues.
> I am not an expert in parquet-thrift so let's wait for @ttim's approval. I'll push this change if there are no objections.

Thanks @gszadovszky! Appreciate the review. I addressed the feedbacks. Could you give it a final pass and help with merging? I can address @ttim's feedbacks, if any, in a follow up diff.
